tests/keras/backend/backend_test.py

```diff
@@ -970,6 +970,107 @@ def step_function(x, states):
         assert_list_pairwise(last_output_list, shape=False)
         assert_list_pairwise(outputs_list, shape=False)
 
+    def test_rnn_output_and_state_masking_independent(self):
+        num_samples = 2
+        num_timesteps = 4
+        state_and_io_size = 5
+
+        def step_function(inputs, states):
+            return inputs, [s + 1 for s in states]
+
+        inputs_vals = np.ones((num_samples, num_timesteps, state_and_io_size))
```

Comment:
I think the `WITH_NP` approach makes sense for the "basic tensor operations" where the numpy implementation represents the reference source of truth. However, for more complicated operations (such as the `rnn`) I think it is bad to test the backends by comparing to the reference backend. If we do that  **we must write exactly this kind of tests with "by hand calculation" to verify _its_ functionality** And when we do, we might just as well run those tests  on the real backends?

Infact, I think it is worrying that there aren't proper input/output test for some methods in the numpy backend. Take the `pool` function for example - it is 50 lines of code that we just take as ground truth, but there are no tests for this code veryfying that it does what we want it to!? (Or am I just not finding those tests?)

Allthough, for completness, it would be nice to have a functioning `rnn` implementation in the numpy backend. However, it is curently nowhere near a proper implementation (it has a different signature and deals with the very special case of a "SimpleRNN") so I don't think that's within the scope of this PR.

Reply:
Actually, what we do right now is, when running travis, we compare:
* theano against numpy
* tensorflow against numpy
* cntk against numpy

those are three different jobs. If one backend is wrong, the corresponding job will fail. If the numpy backend is wrong, the three jobs will fail.

Running some kind of correctness computation is really difficult to do right, because you would need to use random input values to ensure correctness in a majority of cases. But then if you take random values as input, it's very likely that you won't use a paper and a pencil to compute the expected output. You'll likely use numpy. So I think this is our best bet to avoid shallow tests.

After all this, I didn't dig much into the rnn implementation in the numpy backend, so I can't guarantee that it supports your use case. And if it does not, it would surely fall outside the scope of this PR, and we're better off with your implementation of the tests for sure.

After a bit of digging into the git blame, I found that the numpy RNN implemtation was done there: #9557

Maybe @taehoonlee can tell us more about it?

Otherwise I'm fine with the tests that you made if it means changing a lot of code to adapt it to my request. I don't intend to make you do too much extra work since you're already helping a lot with your frequent PRs and discussions :)

Reply:
Thanks for your detailed response @gabrieldemarmiesse . I definitely see how the reference implementation is convenient and improves test coverage 👍  My main point is that if we _only_ rely on "testing by agreement" then _agreement_ is also the only guarantee we get from the tests.

> those are three different jobs. If one backend is wrong, the corresponding job will fail. If the numpy backend is wrong, the three jobs will fail.

But all jobs will run on the same commit, which means if I make a small modification in _all backends_ in the same commit/PR (which is almost what I do here!) then a backward **in**compatible change can go undetected. I understand that this is not likely to happen in practice, but principally it requires the external procedure that you are not allowed to change all backends simultaneously to be safe.

Even if the backends are separate implementations they are still "logically" strongly coupled which means that it is not "completely unlikely" that a bug could be introduced simultaneously in all backends (think e.g. an "innocent PR" aligning variable naming in all of them for improved readability).

It also makes it harder to reason about the code changes; if you have version _X_ and a later version _Y_  of the code (possibly separated by hundreds of commits) and want to verify if a certain function _f_ in _Y_ is backwards compatible with respect to _X_. With input/output unit tests it is a sufficient guarantee that the tests of _f_ in _X_ are still there in _Y_ (assuming good test coverage). To do the same thing with the "agreement tests" you would basically need the full history of Travis logs and check that there are never simultaneous changes in all backends between registered passing tests...

My bottom line is just that (at least for the "more complicated" backend operations) it is worth having the hand calculated tests - for proper "grounding" 😄

Reply:
...It shouldn't be that much work to update the numpy rnn, but let's do it in another PR (and let's keep the hand calculated i/o tests even when its introduced 🙏)

Reply:
Sure thing! I agree that those hand calculated test are important and we'll keep them. You made a very good point.